src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/azuredevops/AzureDevOpsServerPullRequestDecorator.java

```diff
@@ -0,0 +1,438 @@
+        return azureUrl + azureProjectId +
+                "/_apis/git/repositories/" +
+                azureRepositoryName +
+                "/pullRequests/" +
+                pullRequestId +
+                "/threads" +
+                getApiVersion();
+    }
+
+    private String getGitPullRequestStatus(AnalysisDetails analysisDetails) throws IOException {
+        final String GIT_STATUS_CONTEXT_GENRE = "SonarQube";
+        final String GIT_STATUS_CONTEXT_NAME = "QualityGate";
+        final String GIT_STATUS_DESCRIPTION = "SonarQube Gate";
+
+        GitPullRequestStatus status = new GitPullRequestStatus(
+                GitStatusStateMapper.toGitStatusState(analysisDetails.getQualityGateStatus()),
+                GIT_STATUS_DESCRIPTION,
+                new GitStatusContext(GIT_STATUS_CONTEXT_GENRE, GIT_STATUS_CONTEXT_NAME), // "SonarQube/PullRequestDecoration",
+                String.format("%s/dashboard?id=%s&pullRequest=%s", server.getPublicRootUrl(),
+                        URLEncoder.encode(analysisDetails.getAnalysisProjectKey(),
+                                StandardCharsets.UTF_8.name()),
+                        URLEncoder.encode(analysisDetails.getBranchName(),
+                                StandardCharsets.UTF_8.name())
+                )
+        );
+        return new ObjectMapper().writeValueAsString(status);
+    }
+
+    private void sendPost(String apiUrl, String body, String successMessage) throws IOException, MalformedURLException, URISyntaxException {
+        apiUrl = encodeURI(apiUrl);
+        LOGGER.trace(String.format("sendPost: URL: %s ", apiUrl));
+        LOGGER.trace(String.format("sendPost: BODY: %s ", body));
+        HttpPost httpPost = new HttpPost(apiUrl);
+        httpPost.addHeader("Accept", "application/json");
+        httpPost.addHeader("Content-Type", "application/json; charset=utf-8");
+        httpPost.addHeader("Authorization", authorizationHeader);
+        StringEntity entity = new StringEntity(body, StandardCharsets.UTF_8);
+        httpPost.setEntity(entity);
+
+        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+            HttpResponse httpResponse = httpClient.execute(httpPost);
+            if (null != httpResponse && httpResponse.getStatusLine().getStatusCode() != 200) {
+                LOGGER.error("sendPost: " + httpResponse.toString());
+                LOGGER.error("sendPost: " + EntityUtils.toString(httpResponse.getEntity(), StandardCharsets.UTF_8));
+                throw new IllegalStateException("An error was returned in the response from the Azure DevOps server. See the previous log messages for details");
+            } else if (null != httpResponse) {
+                LOGGER.debug("sendPost: " + httpResponse.toString());
+                LOGGER.info("sendPost: " + successMessage);
+            }
+        }
+    }
+
+    private <T> T sendGet(String apiUrl, Class<T> type) throws IOException, MalformedURLException, URISyntaxException {
```

Comment:
The `MalformedURLException` is a subtype of `IOException` so is covered by that `throws` so can be removed.